doc/manual/merging.md

```diff
@@ -0,0 +1,526 @@
+custom merge functions. They are not detailed here yet. For more details, see
+the associated technical document [RFC001][rfc001].
+
+The section describes the behavior and use-cases of merge, by considering the
+following situations:
+
+- [Merging two records without common fields](#simple-merge-(no-common-fields))
+- [Merging records with common fields](#recursive-merge-(with-common-fields))
+- [Merging records with metadata](#merging-record-with-metadata)
+  * [Default values](#default-values)
+  * [Contracts](#contracts)
+  * [Documentation](#documentation)
+- [Recursive overriding](#recursive-overriding)
+
+## Simple merge (no common fields)
+
+Merging two records with no common fields results in a record with the fields
+from both operands. That is, `{foo = 1, bar = "bar"} & {baz =false}` evaluates
+to `{foo = 1, bar = "bar", baz = false}`.
+
+### Specification
+
+Technically, if we write the left operand as:
+
+```
+left = {
+  field_left_1 = value_left_1,
+  ..,
+  field_left_n = value_left_n,
+}
+```
+
+And the right operand as:
+
+```
+right {
+  field_right_1 = value_right_1,
+  ..,
+  field_right_k = value_right_k
+}
+```
+
+Then the merge `left & right` evaluates to the record:
+
+```
+{
+  field_left_1 = value_left_1,
+  ..,
+  field_left_n = value_left_n,
+  field_right_1 = value_right_1,
+  ..,
+  field_right_k = value_right_k
+}
+```
+
+In other terms, `left & right` is the union of `left` and `right`.
+
+### Examples
+
+#### Split
+
+You can split a configuration into subdomains:
+
+```nickel
+// file: server.ncl
+{
+    host_name = "example",
+    host = "example.org",
+    ip_addr = "0.0.0.0",
+}
+
+// file: firewall.ncl
+{
+    enable_firewall = true,
+    open_ports = [23, 80, 443],
+}
+
+// file: network.ncl
+let server = import "server.ncl" in
+let firewall = import "firewall.ncl" in
+server & firewall
+```
+
+This gives:
+
+```nickel
+{
+    host_name = "example",
+    host = "example.org",
+    ip_addr = "0.0.0.0",
+    enable_firewall = true,
+    open_ports = [23, 80, 443],
+}
+```
+
+#### Extension
+
+Given a configuration, you can use merge to add new fields:
+
+```nickel
+// file: safe-network.ncl
+let base = import "network.ncl" in
+base & {use_iptables = true}
+```
+
+## Recursive merge (with common fields)
+
+When the two operands have fields in common, those fields are recursively
+merged. For example:
+
+```nickel
+{
+  top_left = 1,
+  common = {left = "left"}}
+& {
+  top_right = 2,
+  common = {right = "right"}
+}
+```
+
+Evaluates to the record
+
+```nickel
+{
+  top_left = 1,
+  top_right = 2,
+  common = {left = "left", right = "right"}
+}
+```
+
+When one or both of the common fields are not records, the merge will fail
+unless one of the following condition hold:
+
+- They are both of a primitive data type `Num`, `Bool`, `Enum`, or they are null, and they are equal
+- They are physically equal, meaning that they point to the same location in
```

Comment:
To be clear, this pointer equality stuff is not implemented yet (but very soon). On one side, I can see why relying on this may be fragile on the user side. It means that `{x = ["foo"]} & {x = ["foo"]}` doesn't merge, but `let y = ["foo"] in {x = y} & {x = y}` does. On the other hand, I guess as long as you under approximate real equality, it's still better than rejecting everything. In particular in Nickel, where I had problems when applying the same contract several times because any default value that is not primitive (a list, a function, etc.) wouldn't merge with itself). Also it probably helps performance because you just shortcut merging big records if they are actually the same.

But @edolstra may have a point: maybe what we want is to do it nevertheless, to make multiple contract application to never fail, but not document it in the specification, to avoid having people relying on this, and pretending that only equal primitive values can merge.

Reply:
> It means that `{x = ["foo"]} & {x = ["foo"]}` doesn't merge, but `let y = ["foo"] in {x = y} & {x = y}` does.

Oh, yeah that actually sounds pretty bad, breaking referential transparency..

Reply:
So, I just tried to implement this, and recursive records/revertible thunks make the thing even more tricky (in that it doesn't short-cut even when sub-element of the record are the same, which was my original motivation). So let us just get rid of that for now totally, both in the documentation and in the implementation :slightly_smiling_face: